### PR TITLE
[ RFC ] merge portage-stable, coreos-overlay submodules to paths in scripts/

### PR DESCRIPTION
**Pull request by t-lo**
PR is superseded by

* https://github.com/flatcar/scripts/pull/708
* https://github.com/flatcar/scripts/pull/704
* https://github.com/flatcar/scripts/pull/705
* https://github.com/flatcar/scripts/pull/706
* https://github.com/flatcar/scripts/pull/707

The objective and discussion below remains valid and is referenced by above PRs.

This change is a Request for Comment / Proof of Concept to merge both ebuild git submodules into paths while retaining the entire history of both submodules.

The vast majority of commits in this PR is the entire history of https://github.com/flatcar/coreos-overlay and https://github.com/flatcar/portage-stable, merged following the how-to at https://x3ro.de/integrating-a-submodule-into-the-parent-repository/.

To aid reviewing, there is only a single commit on top of the submodules migration. This commit is amended with necessary changes to make the PR work. Please review that one commit only - otherwise Github's web interface will blow up :)
* c5b00016a13e1d4f7a52caae94c137572e3c0498

## Automation
The whole process of merging submodules as paths into the `scripts` repo is automated in a script: https://gist.github.com/t-lo/f403f2f1df5368a46a1454fe08e064f9 .
This script:
1. clones all necessary repos with the version (branch) provided
2.  rewrites paths in coreos-overlay and portage-stable
3. removes submodules in scripts
4. imports the re-written history from 2.
5. clones a pristine version of `scripts` with submodules
6. runs a diff between pristine and merged `scripts`
7. cherry-picks the commit ref above to import build automation changes
8. commits everything

The script only creates local changes and creates a new branch for these changes. After the script ran (and succeeded) the resulting branch may be pushed to create a PR.

## Motivation

Having a single repo simplifies clones, checkouts, and builds. It reduces complexity of existing and future automation. And it immediately enables github action CI builds and tests for PRs against former submodules (which would now be filed against the scripts repo).

## Timeline

In a maintainers call on 2023-04-05 we aligned on the following timeline:
1. After the next releases (https://github.com/flatcar/Flatcar/issues/996 - week of April 10th) we enact a  merge-freeze for coreos-overlay and portage-stable.
2. This submodule merge PRs are re-done with the latest "frozen" state of coreos-overlay and portage-stable. This is straightforward as only the tip commit contains any business logic. This includes:
   a. This here PR (main and then-alpha)
   b. Related PRs to flatcar-3510 (then-stable), flatcar-3549 (then-beta), flatcar-3033 (LTS)
3. the PRs are merged.
4. `scripts` repo is re-tagged so release tags point to the unified scripts. This will make upcoming release notes for successive releases straightforward to generate.
5. related PRs for [github actions](https://github.com/flatcar/scripts/pull/702), [release scripts](https://github.com/flatcar/flatcar-build-scripts/pull/141), and jenkins automation are merged.

## Follow-up work

- [x] Fall-out in the nightly builds / release builds need to be taken care of.
- [ ] Existing automation from the submodule repos - particularly automated PRs - need to be ported to scripts.
- [ ] Update developer documentation and remove references to submodules and the need to use checkout script

### CODE REVIEW

**Comment by t-lo** [comment]
I'm setting this to "Ready for review" even though it probably isn't - this will trigger GH actions build and test, which will give us a firs impression on this PR's feasibility.

**Comment by t-lo** [comment]
Oh, my old friend `emerge-gitclone` causes some trouble once again. Should be easy to fix though. All other tests are green, YAY!

File: bootstrap_sdk [hunk 1]
```diff
@@ -113,50 +113,54 @@ chmod 1777 "${ROOT_OVERLAY}/tmp"
...
 #  This is why we must prevent any updated package ebuilds to "leak" into stage 1, hence we use
 #  "known good" ebuild repo versions outlined above.
 #
-# In special circumstances it may be required to circumvent this and use git
-#  refs of either (or both) portage and overlay. The command line options
-#  --stage1-portage-ref and --stage1-overlay-ref may be used to specify
+# In special circumstances it may be required to circumvent this and use custom paths
+#  for either (or both) portage and overlay. The command line options
+#  --stage1-portage-path and --stage1-overlay-path may be used to specify
 #  a git ref known to work for stage1. In that case the stage1 seed (i.e. the seed SDK)
```
Comment (krnowak) [inline bootstrap_sdk:139]:
```suggestion
#  a repo path known to work for stage1. In that case the stage1 seed (i.e. the seed SDK)
```

**Comment by t-lo** [comment]
@krnowak  Thank you for taking a look!
Yes, workflows are missing - these are on the TODO list (follow-up work).
I agree the PR is hard to review on Github; I'll squash all my changes on top of the submodule migration so we can review a single commit. Should make things a bit easier.

**Comment by t-lo** [comment]
==> https://github.com/flatcar/flatcar-dev-util/pull/15 for the devcontainer test breakage. Will restart tests now.

**Comment by t-lo** [comment]
For some reason the build won't start automatically so I started the action manually: https://github.com/flatcar/scripts/actions/runs/4574599483

**Comment by pothos** [comment]
> This change is only proposed for the main branch. No backports to maintenance branches are planned at this time. The change will organically trickle down to release branches with new releases as these happen.

I would rather avoid this and do it for all branches at once. Otherwise we have to maintain the two ways of building with differences that have to be kept in mind (for quite a long time if this is not in LTS) and it will be harder to backport other build system changes. We would have to migrate the GitHub Actions and Jenkins build/release logic for the main branch and a future Alpha anyway and then doesn't sound like a lot work to backport it to the other branches.

**Comment by pothos** [comment]
We should plan to do the backport in the weeks after the merge. Before the merge we can already test this branch on Jenkins and prepare things to get ready (jenkins-os, flatcar-build-scripts).

**Comment by jepio** [comment]
Re: timeline

> 4. The PR is merged.
> 5. Nightly builds and release build automation for the "main" branch (aka "alpha-next") is switched from submodules to unified scripts. Fallout breakage is fixed.

I don't think it is reasonable to leave work that we know is going to be necessary until after the merge. Parallel night build and release automation can be setup for the branch to ensure that we don't have broken automation after the merge.

**Comment by t-lo** [comment]
> Re: timeline
> 
> > 4. The PR is merged.
> > 5. Nightly builds and release build automation for the "main" branch (aka "alpha-next") is switched from submodules to unified scripts. Fallout breakage is fixed.
> 
> I don't think it is reasonable to leave work that we know is going to be necessary until after the merge. Parallel nightly build and release automation can be setup for the branch to ensure that we don't have broken automation after the merge.

That's a good point. I've taken a stab on this and started a test build on our internal nightly builds infrastructure: http://jenkins.infra.kinvolk.io:8080/job/container/job/sdk/743/cldsv/ to gauge the amount of pain a switch to unified `scripts` would cause.

@pothos you've mentioned release automation is also in need of updating (mainly tagging submodules) - could you please provide a pointer?

**Comment by pothos** [comment]
The release tagging is here:
https://github.com/flatcar/flatcar-build-scripts/blob/master/tag-release
Changelog generation is here: https://github.com/flatcar/flatcar-build-scripts/blob/master/show-changes
The new Alpha branches get created with this script: https://github.com/flatcar/flatcar-build-scripts/blob/master/mirror-repos-branch (Optimization to do is dropping the creation of branches in unused repos)
The rest should be ok :crossed_fingers:

**Comment by t-lo** [comment]
All tests are green on Jenkins CI: (internal infra link): http://jenkins.infra.kinvolk.io:8080/job/container/job/sdk/748/cldsv/

**Comment by pothos** [comment]
> The release tagging is here: https://github.com/flatcar/flatcar-build-scripts/blob/master/tag-release Changelog generation is here: https://github.com/flatcar/flatcar-build-scripts/blob/master/show-changes The new Alpha branches get created with this script: https://github.com/flatcar/flatcar-build-scripts/blob/master/mirror-repos-branch (Optimization to do is dropping the creation of branches in unused repos) The rest should be ok crossed_fingers

Prepared in https://github.com/flatcar/flatcar-build-scripts/pull/141

**Pull request closed without merging**